util/fhir/resourceCreation.ts

```diff
@@ -0,0 +1,152 @@
+import { v4 as uuidv4 } from 'uuid';
+import { getRandomFirstName, getRandomLastName } from '../randomizer';
+import _ from 'lodash';
+import { getResourcePrimaryDates } from './dates';
+import { getResourcePatientReference } from './patient';
+import { getResourceCode } from './codes';
+
+export function createPatientResourceString(birthDate: string): string {
+  const id = uuidv4();
+
+  // NOTE: should add non-binary genders in the future
+  const gender = Math.random() < 0.5 ? 'male' : 'female';
+
+  const pt: fhir4.Patient = {
+    resourceType: 'Patient',
+    id,
+    identifier: [
+      {
+        use: 'usual',
+        system: 'http://example.com/test-id',
+        value: `test-patient-${id}`
+      }
+    ],
+    name: [
+      {
+        family: getRandomLastName(),
+        given: [getRandomFirstName(gender)]
+      }
+    ],
+    gender,
+    birthDate
+  };
+
+  return JSON.stringify(pt, null, 2);
+}
+
+/**
+ * Creates copies of all passed in resources (without references maintained) and gives them
+ * new resource ids. Replaces all patient references to the patient oldId with newId
+ * @param copyResources {fhir4.FhirResource[]} array of fhir resources to be copied
+ * @param oldId {String} a patient id that the copyResources may reference
+ * @param newId {String} a patient id that should replace oldId in references
+ * @returns {fhir4.FhirResource[]} array of new resource copies
+ */
+export function createCopiedResources(
+  copyResources: fhir4.FhirResource[],
+  oldId: string,
+  newId: string
+): fhir4.FhirResource[] {
+  const resources: fhir4.FhirResource[] = copyResources.map(cr => {
+    let resourceString = JSON.stringify(cr);
+    const idRegexp = new RegExp(`Patient/${oldId}`, 'g');
+    resourceString = resourceString.replace(idRegexp, `Patient/${newId}`);
+    const resource: fhir4.FhirResource = JSON.parse(resourceString);
+    resource.id = uuidv4();
+    // Note: this does not update potential cross-resource references, which we may want to support in the future
+    return resource;
+  });
+  return resources;
+}
+
+/**
+ * Creates a copy of the passed in patient object (without references maintained) and updates the
+ * id and identifier as well as creating a new name to differentiate the new patient copy
+ * @param copyPatient {fhir4.Patient} a fhir Patient object to copy
+ * @returns {fhir4.Patient} the new fhir patient copy
+ */
+export function createCopiedPatientResource(copyPatient: fhir4.Patient): fhir4.Patient {
```

Comment:
Even though the name is createCopiedPatientResource, this function creates a copy of a the passed in patient object. Since the return type is fhir4.Patient, should we put this in `patient.ts`? Open to hear the reasoning for keeping this here though!

Reply:
I went with the mantra that anything that's generating a FHIR resource at all goes in resourceCreation, even if that thing is a patient. It's definitely not perfect but I think it might be fine? Idk at this point

Reply:
Ah I see! That works for me